drivers/timer/rk3568/timer.c

```diff
@@ -0,0 +1,207 @@
+    uint32_t load_count1;
+    uint32_t current_value0;
+    uint32_t current_value1;
+    uint32_t control_reg;
+    uint32_t reserved0;
+    uint32_t int_status;
+    uint32_t reserved1;
+} rk3568_timer_regs_t;
+
+static volatile rk3568_timer_regs_t * timestamp_timer;
+static volatile rk3568_timer_regs_t * timeout_timer;
+sddf_channel timestamp_irq;
+sddf_channel timeout_irq;
+
+static uint64_t timeouts[MAX_TIMEOUTS];
+
+static void print_regs(volatile rk3568_timer_regs_t * timer) {
+    sddf_dprintf("TIMER DRIVER|LOG: regs load_count0 : 0x%x\n", timer->load_count0);
+    sddf_dprintf("TIMER DRIVER|LOG: regs load_count1 : 0x%x\n", timer->load_count1);
+    sddf_dprintf("TIMER DRIVER|LOG: regs current_value0: 0x%x\n", timer->current_value0);
+    sddf_dprintf("TIMER DRIVER|LOG: regs current_value1: 0x%x\n", timer->current_value1);
+    sddf_dprintf("TIMER DRIVER|LOG: regs control_reg: 0x%x\n", timer->control_reg);
+    sddf_dprintf("TIMER DRIVER|LOG: regs int_status: 0x%x\n", timer->int_status);
+}
+
+static inline uint64_t get_ticks_in_ns(void)
+{
+    /* the timer value counts down from the load value */
+    uint64_t load_values = timestamp_timer->current_value0;
+    load_values |= (uint64_t)timestamp_timer->current_value1 << 32;
+    uint64_t ticks = UINT64_MAX - load_values;
+
+    /* convert from ticks to nanoseconds */
+    uint64_t value_whole_seconds = ticks / RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ticks = ticks % RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ns = (value_subsecond_ticks * NS_IN_S) / RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_ns = value_whole_seconds * NS_IN_S + value_subsecond_ns;
+
+    //sddf_dprintf("TIMER DRIVER|LOG: get_ticks_in_ns load_values: %lu value_ns: %lu\n", load_values, value_ns);
+    return value_ns;
+}
+
+void set_timeout(uint64_t ns)
+{
+    //sddf_dprintf("TIMER DRIVER|LOG: TIMEOUT SET_TIMEOUT\n");
+    //print_regs(timeout_timer);
+    /* load the timeout timer with ticks to count down from */
+    uint64_t ticks_whole_seconds = (ns / NS_IN_S) * RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t ticks_remainder = (ns % NS_IN_S) * RK3568_TIMER_TICKS_PER_SECOND / NS_IN_S ;
+    uint64_t num_ticks = ticks_whole_seconds + ticks_remainder;
+    uint32_t timeout_ticks_l = num_ticks & 0x00000000ffffffff;
+    uint32_t timeout_ticks_h = num_ticks & 0xffffffff00000000;
+
+    timeout_timer->control_reg = 0x0;
+
+    timeout_timer->load_count0 = timeout_ticks_l;
+    timeout_timer->load_count1 = timeout_ticks_h;
+
+    timeout_timer->control_reg = (RK3568_TIMER_CONTROL_TIMER_ENABLE |
+                                  RK3568_TIMER_CONTROL_MODE_USER |
+                                  RK3568_TIMER_CONTROL_INTERRUPT_ENABLE);
+    //sddf_dprintf("TIMER DRIVER|LOG: set_timeout timeout_ns: %lu ticks: %lu\n", ns, num_ticks);
+    //print_regs(timeout_timer);
+    return;
+}
+
+static void process_timeouts(uint64_t curr_time)
+{
+    for (int i = 0; i < MAX_TIMEOUTS; i++) {
```

Comment:
If it is merged before this, you may want to rebase on the apb_timer commit and use the timeout heap library there. We are getting rid of all of the timers with linear lookup timeouts, although you can skip this if you want since we will also be adding a virtualiser (inc. the heap) to replace all timeout multiplexing currently in the drivers (so changes will come later no matter what). https://github.com/au-ts/sddf/pull/545/files#diff-57bc8a4659346ab05e173e9da3e41a8b3bcffe48b3afe295f79022e52a8042ea

Reply:
Leaving that for later. For now will merge with current way of doing things.